test/pjrt/test_train_pjrt_mnist.py

```diff
@@ -0,0 +1,202 @@
+import args_parse
+
+FLAGS = args_parse.parse_common_options(
+    datadir='/tmp/mnist-data',
+    batch_size=128,
+    momentum=0.5,
+    lr=0.01,
+    target_accuracy=98.0,
+    num_epochs=18)
+
+import os
+import shutil
+import sys
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from torchvision import datasets, transforms
+import torch_xla
+import torch_xla.debug.metrics as met
+import torch_xla.distributed.parallel_loader as pl
+import torch_xla.utils.utils as xu
+import torch_xla.core.xla_model as xm
+import torch_xla.distributed.xla_multiprocessing as xmp
+import torch_xla.test.test_utils as test_utils
+
+
+class MNIST(nn.Module):
+
+  def __init__(self):
+    super(MNIST, self).__init__()
+    self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
+    self.bn1 = nn.BatchNorm2d(10)
+    self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
+    self.bn2 = nn.BatchNorm2d(20)
+    self.fc1 = nn.Linear(320, 50)
+    self.fc2 = nn.Linear(50, 10)
+
+  def forward(self, x):
+    x = F.relu(F.max_pool2d(self.conv1(x), 2))
+    x = self.bn1(x)
+    x = F.relu(F.max_pool2d(self.conv2(x), 2))
+    x = self.bn2(x)
+    x = torch.flatten(x, 1)
+    x = F.relu(self.fc1(x))
+    x = self.fc2(x)
+    return F.log_softmax(x, dim=1)
+
+
+def _train_update(device, x, loss, tracker, writer):
+  test_utils.print_training_update(
+      device,
+      x,
+      loss.item(),
+      tracker.rate(),
+      tracker.global_rate(),
+      summary_writer=writer)
+
+
+def train_mnist(flags, index, state_dict, **kwargs):
+  rank = int(os.getenv('CLOUD_TPU_TASK_ID', 0)) + index
```

Comment:
Should this line be
```
rank = int(os.getenv('CLOUD_TPU_TASK_ID', 0)) * 2 + index
```
on TPU v3 instead (that is, adding `*2` to `CLOUD_TPU_TASK_ID`)? Otherwise, `rank` is one of 0, 1, 2, 3, 4 (**including 4**) on v3-8, which seems like a bug to me. I think it should be one of 0, 1, 2, 3, 4, 5, 6, 7 instead on v3-8.

Similarly for the ImageNet example

Reply:
I think you are right, down below we set 
```
          num_replicas=xm.xrt_world_size(),

```

which I think `xm.xrt_world_size()` should return 8, so our rank should be 0-7. This might explain why we see some accuracy gap on v3-8 using pjrt. @ronghanghu nice catch!

Reply:
Thanks for flagging this line. It is wrong, but actually not for this reason. On TPU v2 and v3, we have 4 processes (`CLOUD_TPU_TASK_ID` 0 through 3) with two visible devices each (two threads with `index` 0 or 1). So on process 0, we have ranks `0 * 2 + 0 = 0` and `0 * 2 + 1 = 1`, and on process 3, we have `3 * 2 + 0 = 6` and `3 * 2 + 1 = 7`. This gives us the range 0 through 7 that we expect.

I didn't fix this line when I added the v4 tests, which have 4 processes with one visible device each (i.e. index is always 0). These will have ranks 0, 2, 4, and 6 but have world size 4. This is just a clunky expression anyway, so I think the right thing to do is update functions in `xla_model` like `xm.get_ordinal()` and `xm.get_local_ordinal()` for PjRt so we don't need to compute the rank like this at all. I see you've already filed a bug to do that (#3692)

Reply:
Thanks @will-cromar! Yeah, I also feel the right thing to do is update functions in xla_model like xm.get_ordinal() and xm.get_local_ordinal().